src/merkle/store/tests.rs

```diff
@@ -647,6 +647,70 @@ fn test_constructors() -> Result<(), MerkleError> {
     Ok(())
 }
 
+#[test]
+fn node_path_should_be_truncated_by_midtier_insert() {
+    let key = 0b11010010_11001100_11001100_11001100_11001100_11001100_11001100_11001100_u64;
+
+    let mut store = MerkleStore::new();
+    let root: Word = EmptySubtreeRoots::empty_hashes(64)[0].into();
+
+    // insert first node - works as expected
+    let depth = 64;
+    let node = [Felt::new(key); WORD_SIZE];
+    let index = NodeIndex::new(depth, key).unwrap();
+    let root = store.set_node(root, index, node).unwrap().root;
+    let result = store.get_node(root, index).unwrap();
+    let path = store.get_path(root, index).unwrap().path;
+    assert_eq!(node, result);
+    assert_eq!(path.depth(), depth);
+    assert!(path.verify(index.value(), result, &root));
+
+    // flip the first bit of the key and insert the second node on a different depth
+    let key = key ^ (1 << 63);
+    let key = key >> 8;
+    let depth = 56;
+    let node = [Felt::new(key); WORD_SIZE];
+    let index = NodeIndex::new(depth, key).unwrap();
+    let root = store.set_node(root, index, node).unwrap().root;
+    let result = store.get_node(root, index).unwrap();
+    let path = store.get_path(root, index).unwrap().path;
+    assert_eq!(node, result);
+    assert_eq!(path.depth(), depth);
+    assert!(path.verify(index.value(), result, &root));
+
+    // attempt to fetch a path of the second node to depth 64
+    // should fail because the previously inserted node will remove its sub-tree from the set
+    let key = key << 8;
+    let index = NodeIndex::new(64, key).unwrap();
+    assert!(store.get_node(root, index).is_err());
+}
+
+#[test]
+fn get_leaf_depth_works() {
+    let key = u64::MAX;
+
+    let mut store = MerkleStore::new();
+    let mut root: Word = EmptySubtreeRoots::empty_hashes(64)[0].into();
+
+    // use the same root to test different paths
+    for depth in (1..=64).rev() {
+        let mask = 64 - depth as u64;
+        let key = (key << mask) >> mask;
```

Comment:
nit: It took me a second to realize what this was doing, one thing that I like to do for expressions like this, is to move them to stand alone functions, so that they can be unit tested, and give a name to what is being done, for example:

```rust
fn drop_high_bits(number: u64, bit_pos: usize) -> u64 {
   (number << bit_pos) >> bit_pos
}
```

alternatively this could also be a comment

Reply:
Agreed! Would be good to do this for clarity.

Reply:
I rewrote that test, it was indeed quite confusing. I'll add now a variant to flip + decrement the depth. Let's keep the discussion centralized here

https://github.com/0xPolygonMiden/crypto/pull/125#discussion_r1158687618